### PR TITLE
_7zz: 24.05 → 24.06

### PR DESCRIPTION
**Pull request by jtojnar**
## Description of changes

https://sourceforge.net/p/sevenzip/discussion/45797/thread/bbe6861abe/

Depends on https://github.com/NixOS/nixpkgs/pull/318589

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by jtojnar** [comment]
I had `_7zz.unfree` fail to build a few times:

```
Running phase: buildPhase
build flags: -j8 SHELL=/nix/store/306znyj77fv49kwnkpxmb0j2znqpa8bj-bash-5.2p26/bin/bash CC=cc CXX=c++ MY_ASM=uasm
mkdir -p b/g_x64
uasm -nologo -elf64 -DABI_LINUX -Fob/g_x64/ ../../../../Asm/x86/LzmaDecOpt.asm
cc      -O2 -c -Werror -Wall -Wextra -Waddress -Waggressive-loop-optimizations -Wattributes -Wcast-align -Wcomment -Wdiv-by-zero -Wformat-contains-nul -Winit-self -Wint-to-pointer-cast -Wunused -Wunused-macros  -Wbool-compare -Wduplicated-cond  -Waddress-of-packed-member -Wbool-operation -Wcast-align=strict -Wconversion -Wdangling-else -Wduplicated-branches -Wimplicit-fallthrough=5 -Wint-in-bool-context -Wmaybe-uninitialized -Wmisleading-indentation -Wmissing-attributes -DNDEBUG -D_REENTRANT -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -fPIC   -o b/g_x64/7zBuf2.o ../../../../C/7zBuf2.c
cc      -O2 -c -Werror -Wall -Wextra -Waddress -Waggressive-loop-optimizations -Wattributes -Wcast-align -Wcomment -Wdiv-by-zero -Wformat-contains-nul -Winit-self -Wint-to-pointer-cast -Wunused -Wunused-macros  -Wbool-compare -Wduplicated-cond  -Waddress-of-packed-member -Wbool-operation -Wcast-align=strict -Wconversion -Wdangling-else -Wduplicated-branches -Wimplicit-fallthrough=5 -Wint-in-bool-context -Wmaybe-uninitialized -Wmisleading-indentation -Wmissing-attributes -DNDEBUG -D_REENTRANT -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -fPIC   -o b/g_x64/7zCrc.o ../../../../C/7zCrc.c
uasm -nologo -elf64 -DABI_LINUX -Fob/g_x64/ ../../../../Asm/x86/7zCrcOpt.asm
cc      -O2 -c -Werror -Wall -Wextra -Waddress -Waggressive-loop-optimizations -Wattributes -Wcast-align -Wcomment -Wdiv-by-zero -Wformat-contains-nul -Winit-self -Wint-to-pointer-cast -Wunused -Wunused-macros  -Wbool-compare -Wduplicated-cond  -Waddress-of-packed-member -Wbool-operation -Wcast-align=strict -Wconversion -Wdangling-else -Wduplicated-branches -Wimplicit-fallthrough=5 -Wint-in-bool-context -Wmaybe-uninitialized -Wmisleading-indentation -Wmissing-attributes -DNDEBUG -D_REENTRANT -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -fPIC   -o b/g_x64/7zStream.o ../../../../C/7zStream.c
Fatal error A1106: Cannot open file: "b/g_x64/LzmaDecOpt.o" [ENOENT]
cc      -O2 -c -Werror -Wall -Wextra -Waddress -Waggressive-loop-optimizations -Wattributes -Wcast-align -Wcomment -Wdiv-by-zero -Wformat-contains-nul -Winit-self -Wint-to-pointer-cast -Wunused -Wunused-macros  -Wbool-compare -Wduplicated-cond  -Waddress-of-packed-member -Wbool-operation -Wcast-align=strict -Wconversion -Wdangling-else -Wduplicated-branches -Wimplicit-fallthrough=5 -Wint-in-bool-context -Wmaybe-uninitialized -Wmisleading-indentation -Wmissing-attributes -DNDEBUG -D_REENTRANT -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -fPIC   -o b/g_x64/Aes.o ../../../../C/Aes.c
make: *** [../../7zip_gcc.mak:1307: b/g_x64/LzmaDecOpt.o] Error 1
make: *** Waiting for unfinished jobs....
7zCrcOpt.asm: 258 lines, 3 passes, 1997 ms, 0 warnings, 0 errors
```

But none of those files were modified and it builds now. Maybe some missing dependency revealed by parallel building.

File: pkgs/tools/archivers/7zz/update.sh [hunk 1]
```diff
@@ -14,37 +11,7 @@ if [[ "$OLD_VERSION" == "$NEW_VERSION" ]]; then
...
-
-echo "updating free hash"
-sed -i "s@free = \"$OLD_FREE_HASH\";@free = \"$NEW_FREE_HASH\";@" "$DRV_DIR/default.nix"
-echo "updating unfree hash"
-sed -i "s@unfree = \"$OLD_UNFREE_HASH\";@unfree = \"$NEW_UNFREE_HASH\";@" "$DRV_DIR/default.nix"
-
-echo "done"
+update-source-version _7zz "$NEW_VERSION"
+# Reset version or update will not be performed.
+update-source-version _7zz.unfree 0 "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
```
Comment (eclairevoyant) [inline pkgs/tools/archivers/7zz/update.sh:16]: rest lgtm, but I didn't understand this line?

File: pkgs/tools/archivers/7zz/default.nix [hunk 2]
```diff
@@ -117,6 +119,8 @@ stdenv.mkDerivation (finalAttrs: {
       package = finalAttrs.finalPackage;
       command = "7zz --help";
     };
+    # used by update script
+    unfree = _7zz.override { enableUnfree = true; };
```
Comment (eclairevoyant) [inline pkgs/tools/archivers/7zz/default.nix:123]: also, this is probably fine for the update script, but we should probably note that `_7zz` should never be used anywhere else in the file

File: pkgs/tools/archivers/7zz/update.sh (hunk 1 above)
Comment (jtojnar) [reply pkgs/tools/archivers/7zz/update.sh:16]: u-s-v will not do anything if it detects the version is already up to date.

Comment (jtojnar) [reply pkgs/tools/archivers/7zz/update.sh:16]: Opened https://github.com/NixOS/nixpkgs/pull/318589

File: pkgs/tools/archivers/7zz/default.nix (hunk 2 above)
Comment (jtojnar) [reply pkgs/tools/archivers/7zz/default.nix]: Resolved by including an older commit of mine that exposes the value in top-level.

**Comment by jtojnar** [comment]
Rebased since I merged the u-s-v change-.

**Comment by Lyndeno** [comment]
Result of `nixpkgs-review pr 318506` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>kdePackages.kajongg</li>
    <li>kdePackages.kajongg.debug</li>
    <li>kdePackages.kajongg.dev</li>
    <li>kdePackages.kajongg.devtools</li>
  </ul>
</details>
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>_7zz-rar</li>
    <li>nexusmods-app-unfree</li>
    <li>peazip</li>
  </ul>
</details>
<details>
  <summary>251 packages built:</summary>
  <ul>
    <li>_7zz</li>
    <li>adminer</li>
    <li>bookstack</li>
    <li>composer-require-checker</li>
    <li>davis</li>
    <li>firefly-iii</li>
    <li>flarum</li>
    <li>grocy</li>
    <li>kdePackages.bomber</li>
    <li>kdePackages.bomber.debug</li>
    <li>kdePackages.bomber.dev</li>
    <li>kdePackages.bomber.devtools</li>
    <li>kdePackages.bovo</li>
    <li>kdePackages.bovo.debug</li>
    <li>kdePackages.bovo.dev</li>
    <li>kdePackages.bovo.devtools</li>
    <li>kdePackages.granatier</li>
    <li>kdePackages.granatier.debug</li>
    <li>kdePackages.granatier.dev</li>
    <li>kdePackages.granatier.devtools</li>
    <li>kdePackages.kapman</li>
    <li>kdePackages.kapman.debug</li>
    <li>kdePackages.kapman.dev</li>
    <li>kdePackages.kapman.devtools</li>
    <li>kdePackages.katomic</li>
    <li>kdePackages.katomic.debug</li>
    <li>kdePackages.katomic.dev</li>
    <li>kdePackages.katomic.devtools</li>
    <li>kdePackages.kblackbox</li>
    <li>kdePackages.kblackbox.debug</li>
    <li>kdePackages.kblackbox.dev</li>
    <li>kdePackages.kblackbox.devtools</li>
    <li>kdePackages.kblocks</li>
    <li>kdePackages.kblocks.debug</li>
    <li>kdePackages.kblocks.dev</li>
    <li>kdePackages.kblocks.devtools</li>
    <li>kdePackages.kbounce</li>
    <li>kdePackages.kbounce.debug</li>
    <li>kdePackages.kbounce.dev</li>
    <li>kdePackages.kbounce.devtools</li>
    <li>kdePackages.kbreakout</li>
    <li>kdePackages.kbreakout.debug</li>
    <li>kdePackages.kbreakout.dev</li>
    <li>kdePackages.kbreakout.devtools</li>
    <li>kdePackages.kdiamond</li>
    <li>kdePackages.kdiamond.debug</li>
    <li>kdePackages.kdiamond.dev</li>
    <li>kdePackages.kdiamond.devtools</li>
    <li>kdePackages.kfourinline</li>
    <li>kdePackages.kfourinline.debug</li>
    <li>kdePackages.kfourinline.dev</li>
    <li>kdePackages.kfourinline.devtools</li>
    <li>kdePackages.kgoldrunner</li>
    <li>kdePackages.kgoldrunner.debug</li>
    <li>kdePackages.kgoldrunner.dev</li>
    <li>kdePackages.kgoldrunner.devtools</li>
    <li>kdePackages.kigo</li>
    <li>kdePackages.kigo.debug</li>
    <li>kdePackages.kigo.dev</li>
    <li>kdePackages.kigo.devtools</li>
    <li>kdePackages.killbots</li>
    <li>kdePackages.killbots.debug</li>
    <li>kdePackages.killbots.dev</li>
    <li>kdePackages.killbots.devtools</li>
    <li>kdePackages.kiriki</li>
    <li>kdePackages.kiriki.debug</li>
    <li>kdePackages.kiriki.dev</li>
    <li>kdePackages.kiriki.devtools</li>
    <li>kdePackages.kjumpingcube</li>
    <li>kdePackages.kjumpingcube.debug</li>
    <li>kdePackages.kjumpingcube.dev</li>
    <li>kdePackages.kjumpingcube.devtools</li>
    <li>kdePackages.klickety</li>
    <li>kdePackages.klickety.debug</li>
    <li>kdePackages.klickety.dev</li>
    <li>kdePackages.klickety.devtools</li>
    <li>kdePackages.klines</li>
    <li>kdePackages.klines.debug</li>
    <li>kdePackages.klines.dev</li>
    <li>kdePackages.klines.devtools</li>
    <li>kdePackages.kmahjongg</li>
    <li>kdePackages.kmahjongg.debug</li>
    <li>kdePackages.kmahjongg.dev</li>
    <li>kdePackages.kmahjongg.devtools</li>
    <li>kdePackages.kmines</li>
    <li>kdePackages.kmines.debug</li>
    <li>kdePackages.kmines.dev</li>
    <li>kdePackages.kmines.devtools</li>
    <li>kdePackages.knavalbattle</li>
    <li>kdePackages.knavalbattle.debug</li>
    <li>kdePackages.knavalbattle.dev</li>
    <li>kdePackages.knavalbattle.devtools</li>
    <li>kdePackages.knetwalk</li>
    <li>kdePackages.knetwalk.debug</li>
    <li>kdePackages.knetwalk.dev</li>
    <li>kdePackages.knetwalk.devtools</li>
    <li>kdePackages.knights</li>
    <li>kdePackages.knights.debug</li>
    <li>kdePackages.knights.dev</li>
    <li>kdePackages.knights.devtools</li>
    <li>kdePackages.kolf</li>
    <li>kdePackages.kolf.debug</li>
    <li>kdePackages.kolf.dev</li>
    <li>kdePackages.kolf.devtools</li>
    <li>kdePackages.kollision</li>
    <li>kdePackages.kollision.debug</li>
    <li>kdePackages.kollision.dev</li>
    <li>kdePackages.kollision.devtools</li>
    <li>kdePackages.konquest</li>
    <li>kdePackages.konquest.debug</li>
    <li>kdePackages.konquest.dev</li>
    <li>kdePackages.konquest.devtools</li>
    <li>kdePackages.kpat</li>
    <li>kdePackages.kpat.debug</li>
    <li>kdePackages.kpat.dev</li>
    <li>kdePackages.kpat.devtools</li>
    <li>kdePackages.kreversi</li>
    <li>kdePackages.kreversi.debug</li>
    <li>kdePackages.kreversi.dev</li>
    <li>kdePackages.kreversi.devtools</li>
    <li>kdePackages.kshisen</li>
    <li>kdePackages.kshisen.debug</li>
    <li>kdePackages.kshisen.dev</li>
    <li>kdePackages.kshisen.devtools</li>
    <li>kdePackages.ksirk</li>
    <li>kdePackages.ksirk.debug</li>
    <li>kdePackages.ksirk.dev</li>
    <li>kdePackages.ksirk.devtools</li>
    <li>kdePackages.ksnakeduel</li>
    <li>kdePackages.ksnakeduel.debug</li>
    <li>kdePackages.ksnakeduel.dev</li>
    <li>kdePackages.ksnakeduel.devtools</li>
    <li>kdePackages.kspaceduel</li>
    <li>kdePackages.kspaceduel.debug</li>
    <li>kdePackages.kspaceduel.dev</li>
    <li>kdePackages.kspaceduel.devtools</li>
    <li>kdePackages.ksquares</li>
    <li>kdePackages.ksquares.debug</li>
    <li>kdePackages.ksquares.dev</li>
    <li>kdePackages.ksquares.devtools</li>
    <li>kdePackages.ksudoku</li>
    <li>kdePackages.ksudoku.debug</li>
    <li>kdePackages.ksudoku.dev</li>
    <li>kdePackages.ksudoku.devtools</li>
    <li>kdePackages.ktuberling</li>
    <li>kdePackages.ktuberling.debug</li>
    <li>kdePackages.ktuberling.dev</li>
    <li>kdePackages.ktuberling.devtools</li>
    <li>kdePackages.kubrick</li>
    <li>kdePackages.kubrick.debug</li>
    <li>kdePackages.kubrick.dev</li>
    <li>kdePackages.kubrick.devtools</li>
    <li>kdePackages.libkdegames</li>
    <li>kdePackages.libkdegames.debug</li>
    <li>kdePackages.libkdegames.dev</li>
    <li>kdePackages.libkdegames.devtools</li>
    <li>kdePackages.libkmahjongg</li>
    <li>kdePackages.libkmahjongg.debug</li>
    <li>kdePackages.libkmahjongg.dev</li>
    <li>kdePackages.libkmahjongg.devtools</li>
    <li>kdePackages.lskat</li>
    <li>kdePackages.lskat.debug</li>
    <li>kdePackages.lskat.dev</li>
    <li>kdePackages.lskat.devtools</li>
    <li>kdePackages.palapeli</li>
    <li>kdePackages.palapeli.debug</li>
    <li>kdePackages.palapeli.dev</li>
    <li>kdePackages.palapeli.devtools</li>
    <li>kdePackages.picmi</li>
    <li>kdePackages.picmi.debug</li>
    <li>kdePackages.picmi.dev</li>
    <li>kdePackages.picmi.devtools</li>
    <li>kdePackages.skladnik</li>
    <li>kdePackages.skladnik.debug</li>
    <li>kdePackages.skladnik.dev</li>
    <li>kdePackages.skladnik.devtools</li>
    <li>laravel</li>
    <li>librenms</li>
    <li>movim</li>
    <li>n98-magerun</li>
    <li>n98-magerun2</li>
    <li>nexusmods-app</li>
    <li>p7zip-rar</li>
    <li>p7zip-rar.doc</li>
    <li>p7zip-rar.lib</li>
    <li>p7zip-rar.man</li>
    <li>paratest</li>
    <li>pdepend</li>
    <li>pest</li>
    <li>phel</li>
    <li>php81Packages.box (php82Packages.box ,php83Packages.box)</li>
    <li>php81Packages.castor</li>
    <li>php81Packages.composer</li>
    <li>php81Packages.composer-local-repo-plugin</li>
    <li>php81Packages.cyclonedx-php-composer</li>
    <li>php81Packages.deployer</li>
    <li>php81Packages.grumphp</li>
    <li>php81Packages.phan</li>
    <li>php81Packages.phing</li>
    <li>php81Packages.phive</li>
    <li>php81Packages.php-codesniffer</li>
    <li>php81Packages.php-cs-fixer</li>
    <li>php81Packages.php-parallel-lint</li>
    <li>php81Packages.phpinsights</li>
    <li>php81Packages.phpmd</li>
    <li>php81Packages.phpstan</li>
    <li>php81Packages.psalm</li>
    <li>php81Packages.psysh</li>
    <li>php82Packages.castor</li>
    <li>php82Packages.composer</li>
    <li>php82Packages.composer-local-repo-plugin</li>
    <li>php82Packages.cyclonedx-php-composer</li>
    <li>php82Packages.deployer</li>
    <li>php82Packages.grumphp</li>
    <li>php82Packages.phan</li>
    <li>php82Packages.phing</li>
    <li>php82Packages.phive</li>
    <li>php82Packages.php-codesniffer</li>
    <li>php82Packages.php-cs-fixer</li>
    <li>php82Packages.php-parallel-lint</li>
    <li>php82Packages.phpinsights</li>
    <li>php82Packages.phpmd</li>
    <li>php82Packages.phpstan</li>
    <li>php82Packages.psalm</li>
    <li>php82Packages.psysh</li>
    <li>php83Packages.castor</li>
    <li>php83Packages.composer</li>
    <li>php83Packages.composer-local-repo-plugin</li>
    <li>php83Packages.cyclonedx-php-composer</li>
    <li>php83Packages.deployer</li>
    <li>php83Packages.grumphp</li>
    <li>php83Packages.phan</li>
    <li>php83Packages.phing</li>
    <li>php83Packages.phive</li>
    <li>php83Packages.php-codesniffer</li>
    <li>php83Packages.php-cs-fixer</li>
    <li>php83Packages.php-parallel-lint</li>
    <li>php83Packages.phpinsights</li>
    <li>php83Packages.phpmd</li>
    <li>php83Packages.phpstan</li>
    <li>php83Packages.psalm</li>
    <li>php83Packages.psysh</li>
    <li>phpactor</li>
    <li>phpdocumentor</li>
    <li>phpunit</li>
    <li>pixelfed</li>
    <li>roave-backward-compatibility-check</li>
    <li>robo</li>
    <li>simplesamlphp</li>
    <li>snipe-it</li>
    <li>vimPlugins.phpactor</li>
  </ul>
</details>

**Review by eclairevoyant — APPROVED** [review]
kajongg and peazip are broken already, and the other two are unfree. lgtm

will wait for CI since some darwin packages use the 7zz setup hook

**Comment by h7x4** [comment]
@eclairevoyant darwin pipeline succeeded

File: pkgs/tools/archivers/7zz/update.sh [hunk 3]
```diff
@@ -14,37 +11,5 @@ if [[ "$OLD_VERSION" == "$NEW_VERSION" ]]; then
...
-sed -i "s/version = \"$OLD_VERSION\";/version = \"$NEW_VERSION\";/" "$DRV_DIR/default.nix"
-
-echo "updating free hash"
-sed -i "s@free = \"$OLD_FREE_HASH\";@free = \"$NEW_FREE_HASH\";@" "$DRV_DIR/default.nix"
-echo "updating unfree hash"
-sed -i "s@unfree = \"$OLD_UNFREE_HASH\";@unfree = \"$NEW_UNFREE_HASH\";@" "$DRV_DIR/default.nix"
-
-echo "done"
+update-source-version _7zz "$NEW_VERSION"
+update-source-version _7zz-rar "$NEW_VERSION" --ignore-same-version
```
Comment (06kellyjac) [inline pkgs/tools/archivers/7zz/update.sh:15]: oh nice

**Pull request merged**